### ISSUE TITLE
RC2 Upgrade Eager Loading throw an exception

### ISSUE THREAD

**Issue by XcoderTeam** [aspnet/EntityFramework#5571]
I upgrade my project from RC1 to RC2 found there is an issue in my query.
In RC1 it work fine, anything wrong with my code?

             var query = _talentRepository.Table;

            if (!string.IsNullOrEmpty(location))
                query = query.Where(t => t.Location.Contains(location));

            query = query.Where(t => t.ResumeSearchable == resumeSearchable);
            
            query = query.Include(t => t.User).ThenInclude(t => t.Skills);
            query = query.Include(t => t.User).ThenInclude(t => t.Websites);
            query = query.Include(t => t.User).ThenInclude(t => t.Languages);
            query = query.Include(t => t.User).ThenInclude(t => t.Experiences);

            if (!string.IsNullOrEmpty(keyword))
            {
                query = query.Where(t => t.User.Name.Contains(keyword) |
                                         t.ShortBio.Contains(keyword) |
                                         t.User.Skills.Any(s => s.Name.Contains(keyword)) |
                                         t.User.Languages.Any(s => s.Name.Contains(keyword)) |
                                         t.User.Experiences.Any(s => s.PositionTitle.Contains(keyword))
                                   );

            }
            return query.ToList();

ApplicationUser.cs

    public class ApplicationUser : IdentityUser<int>
    { 
        [ForeignKey("UserId")]
        public virtual ICollection<UserWebsite> Websites { get; set; }
        [ForeignKey("UserId")]
        public virtual ICollection<UserSkill> Skills { get; set; }
        [ForeignKey("UserId")]
        public virtual ICollection<UserLanguage> Languages { get; set; }
        [ForeignKey("UserId")]
        public virtual ICollection<UserExperience> Experiences { get; set; }
        [ForeignKey("UserId")]
        public virtual ICollection<UserEducation> Educations { get; set; }

        public virtual ICollection<JobPost> JobPost { get; set; }
    }

Exception
System.Data.SqlClient.SqlException was unhandled by user code
  Class=16
  HResult=-2146232060
  LineNumber=17
  Message=Invalid column name 'UserId0'.
Invalid column name 'UserId1'.
Invalid column name 'UserId2'.
  Number=207
  Procedure=""
  Server=DESKTOP-8BU9T6H
  Source=Core .Net SqlClient Data Provider
  State=1
  StackTrace:
       at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
       at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
       at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
       at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
       at System.Data.SqlClient.SqlDataReader.get_MetaData()
       at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
       at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
       at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method)
       at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, String executeMethod, IReadOnlyDictionary`2 parameterValues, Boolean openConnection, Boolean closeConnection)
       at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues, Boolean manageConnection)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable.Enumerator.MoveNext()
       at Microsoft.EntityFrameworkCore.Query.Internal.IncludeCollectionIterator.<GetRelatedValues>d__4.MoveNext()
       at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
       at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.Include(QueryContext queryContext, Object entity, IReadOnlyList`1 navigationPath, IReadOnlyList`1 relatedEntitiesLoaders, Int32 currentNavigationIndex, Boolean queryStateManager)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.<>c__DisplayClass11_0.<Include>b__0(EntityLoadInfo eli)
       at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
       at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.Include(QueryContext queryContext, Object entity, IReadOnlyList`1 navigationPath, IReadOnlyList`1 relatedEntitiesLoaders, Int32 currentNavigationIndex, Boolean queryStateManager)
       at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.Include(QueryContext queryContext, Object entity, IReadOnlyList`1 navigationPath, IReadOnlyList`1 relatedEntitiesLoaders, Boolean queryStateManager)
       at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_Include>d__30`1.MoveNext()
       at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_Include>d__30`1.MoveNext()
       at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_Include>d__30`1.MoveNext()
       at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_Include>d__30`1.MoveNext()
       at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.<_TrackEntities>d__15`2.MoveNext()
       at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
       at Job.Core.PagedList`1..ctor(IQueryable`1 source, Int32 pageIndex, Int32 pageSize) in C:\+Beta\Job\src\Job.Core\PagedList.cs:line 16
       at Job.Services.Account.TalentService.GetAllTalents(Int32 pageIndex, Int32 pageSize, String keyword, String location, Boolean resumeSearchable) in C:\+Beta\Job\src\Job.Services\Account\TalentService.cs:line 93
       at Job.Controllers.Api.TalentController.Get(TalentListModel input) in C:\+Beta\Job\src\Job\Controllers\Api\TalentController.cs:line 93
       at lambda_method(Closure , Object , Object[] )
       at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionAsync>d__6.MoveNext()
  InnerException:

**Comment by XcoderTeam** [comment]
Hi EF Team,

from the profiler capture
`exec sp_executesql N'SELECT [u3].[Id], [u3].[Category], [u3].[Url], [u3].[UserId]
FROM [UserWebsite] AS [u3]
INNER JOIN (
    SELECT DISTINCT [t1].*
    FROM (
        SELECT [u].[UserId], [u2].[UserId] AS [c0]
        FROM [Talent] AS [t]
        INNER JOIN [User] AS [u] ON [t].[UserId] = [u].[UserId]
        INNER JOIN [User] AS [u2] ON [t].[UserId] = [u2].[UserId]
        WHERE [t].[ResumeSearchable] = @__resumeSearchable_0
        ORDER BY [u].[UserId], [u2].[UserId]
        OFFSET @__p_1 ROWS FETCH NEXT @__p_2 ROWS ONLY
    ) AS [t1]
) AS [u20] ON [u3].[UserId] = [u20].[UserId]
ORDER BY [u20].[UserId], [u20].[UserId]',N'@__resumeSearchable_0 bit,@__p_1 int,@__p_2 int',@__resumeSearchable_0=1,@__p_1=0,@__p_2=10`

It generate wrong order by, what is the work around? since RC1 it work upgrading to RC2 this is throw an exception

**Comment by XcoderTeam** [comment]
Hi @maumar 

Any update on this issue?

**Comment by maumar** [comment]
@XcoderTeam not yet, this will be my next bug to look at, will post update when I'm done with the investigation

**Comment by maumar** [comment]
@XcoderTeam can you provide code for:
- Table entity,
- OnModelCreating method of the DbContext,
- UserWebsite, UserSkil, UserLanguage, UserExperience, UserEducation

if they look something more complicated than:

```cs
    public class UserSkill
    {
        public int Id { get; set; }

        public int? UserId { get; set; }

        public string Name { get; set; }
    }
```

**Comment by XcoderTeam** [comment]
@maumar 

`public class Talent : BaseEntity
	{
        public int UserId { get; set; }
        public virtual ApplicationUser User { get; set; }
	}
	
	public class UserSkill : BaseEntity
    {
        public int UserId { get; set; }
        public string Name { get; set; }
        public int Experience { get; set; }
    }
	public class UserLanguage : BaseEntity
    {
        public int UserId { get; set; }
        public string Name { get; set; }
        public LanguageLevel Level { get; set; }
    }
	public class ApplicationUser : IdentityUser<int>
    {
        [ForeignKey("UserId")]
        public virtual ICollection<UserSkill> Skills { get; set; }
        [ForeignKey("UserId")]
        public virtual ICollection<UserLanguage> Languages { get; set; }
    }
	
	protected override void OnModelCreating(ModelBuilder builder)
    {
            base.OnModelCreating(builder);

            builder.Entity<ApplicationUser>(u => {
                u.ToTable("User");

                u.Property(acc => acc.Id).HasColumnName("UserId");
            });
			
            builder.Entity<Talent>(t =>
            {
                t.Property(tal => tal.Id).UseSqlServerIdentityColumn();

                t.Property(tal => tal.UserId).IsRequired();
            });
			
            builder.Entity<UserSkill>(s =>
            {
                s.Property(ski => ski.Id).UseSqlServerIdentityColumn();
                s.Property(ski => ski.UserId).IsRequired();
                s.Property(ski => ski.Name).HasMaxLength(50);
            });
            builder.Entity<UserLanguage>(l =>
            {
                l.Property(lan => lan.Id).UseSqlServerIdentityColumn();
                l.Property(lan => lan.UserId).IsRequired();
                l.Property(lan => lan.Name).HasMaxLength(50);
                l.Property(lan => lan.Level).IsRequired();
            });
    }`

**Comment by maumar** [comment]
@XcoderTeam any chance you can provide full repro for this issue? I'm still unable to reproduce it on my end, not sure what I'm missing. Queries I see look as follows:

```
SELECT [l].[Id], [l].[Level], [l].[Name], [l].[UserId]
FROM [Language] AS [l]
INNER JOIN (
    SELECT DISTINCT [u].[UserId], [u1].[UserId] AS [UserId0]
    FROM [Talents] AS [t]
    INNER JOIN [User] AS [t.User] ON [t].[UserId] = [t.User].[UserId]
    INNER JOIN [User] AS [t.User0] ON [t].[UserId] = [t.User0].[UserId]
    INNER JOIN [User] AS [t.User1] ON [t].[UserId] = [t.User1].[UserId]
    INNER JOIN [User] AS [u] ON [t].[UserId] = [u].[UserId]
    INNER JOIN [User] AS [u1] ON [t].[UserId] = [u1].[UserId]
    WHERE ([t.User].[Name] LIKE (N''%'' + @__keyword_0) + N''%'' OR EXISTS (
        SELECT 1
        FROM [Skills] AS [s]
        WHERE [s].[Name] LIKE (N''%'' + @__keyword_1) + N''%'' AND ([t.User1].[UserId] = [s].[UserId]))) OR EXISTS (
        SELECT 1
        FROM [Language] AS [s0]
        WHERE [s0].[Name] LIKE (N''%'' + @__keyword_2) + N''%'' AND ([t.User0].[UserId] = [s0].[UserId]))
) AS [u10] ON [l].[UserId] = [u10].[UserId0]
ORDER BY [u10].[UserId], [u10].[UserId0]',N'@__keyword_0 nvarchar(4000),@__keyword_1 nvarchar(4000),@__keyword_2 nvarchar(4000)',@__keyword_0=N'Foo',@__keyword_1=N'Foo',@__keyword_2=N'Foo'
```

you can send it directly to my email: [my_github_nickname] [at] micosoft [doT] com

**Comment by maumar** [comment]
@XcoderTeam thanks, I was able to reproduce the issue with the information you provided

**Comment by maumar** [comment]
fixed in 0469dab5be86513197eac95e7aae598a7af127af
